mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/persistence/PersistableState.kt

```diff
@@ -0,0 +1,246 @@
+
+            val branchReference = BranchReference(repositoryId, branchName)
+            var modules = listOf<AbstractModule>()
+            var bindings: Iterable<IBinding>? = null
+            ActiveMpsProjectInjector.runMpsReadAction { repository ->
+                logger.debug { "Restoring SModules." }
+                modules = moduleIds.map {
+                    val id = PersistenceFacade.getInstance().createModuleId(it)
+                    val module = repository.getModule(id)
+                    requireNotNull(module) { "Could not restore module from ID ($id)." }
+                    require(module is AbstractModule) { "Module ($module) is not an AbstractModule." }
+                    module
+                }
+                logger.debug { "${modules.count()} SModules are restored." }
+
+                logger.debug { "Recreating Bindings." }
+                bindings = syncService.rebindModules(client, branchReference, initialVersion, modules)
+            }
+
+            if (bindings == null) {
+                throw IllegalStateException("Rebinding modules failed, thus PersistableState is not restored.")
+            }
+
+            logger.debug { "Bindings are recreated, now activating them." }
+            bindings!!.forEach(IBinding::activate)
+            logger.debug { "Bindings are activated." }
+            return RestoredStateContext(client, branchReference, modules)
+        } catch (t: Throwable) {
+            val message =
+                "Error occurred, while restoring persisted state. Connection to model server, bindings and synchronization cache might not be established and activated. Please check logs for details."
+            notifier.notifyAndLogError(message, t, logger)
+
+            close(client, MpsToModelixMap)
+
+            return null
+        }
+    }
+
+    private fun close(client: ModelClientV2?, cache: MpsToModelixMap) {
+        client?.let {
+            it.close()
+            val message = "Disconnected from server: ${it.baseUrl}"
+            notifier.notifyAndLogInfo(message, logger)
+        }
+        cache.clear()
+    }
+}
+
+/**
+ * Some context after the [PersistableState] has restored the sync plugin's internal state.
+ *
+ * @property modelClient the client that is connected to the model server
+ * @property branchReference the branch in the repository to which we are connected
+ * @property modules the MPS modules for which we established the bindings
+ */
+@UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
+data class RestoredStateContext(
+    val modelClient: ModelClientV2,
+    val branchReference: BranchReference,
+    val modules: List<AbstractModule>,
+)
+
+@UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
+internal class PersistableStateSerializer : KSerializer<PersistableState> {
```

Comment:
Why is the `PersistableStateSerializer` needed? `PersistableState` should be serializable to JSON without a custom serializer.